### PR TITLE
Depth First Search in Python3

### LINKED ISSUE #492: Depth First Search in Python3
This is an open issue for the Depth First Search algorithm implementation in Python3. If you wish to contribute, just make some noise and we will assign you.

### PR DESCRIPTION
**Pull request by vikumkbv**
**Hi!**  
Thanks for considering for contributing to this ever-growing list of algorithm and data structure implementations.
Your contribution is valuable. Use the follow template for creating PRs. (Delete this part)
---

**Fixes #492 **

By submitting this pull request I confirm I've read and complied with the below declarations.

-   [&#10004; ] I have read the [Contribution guidelines](https://github.com/FOSS-UCSC/FOSSALGO/blob/master/CONTRIBUTING.md) 
and I am confident that my PR reflects them.

-   [&#10004; ] I have followed the [coding guidelines](https://github.com/FOSS-UCSC/FOSSALGO/blob/master/CONTRIBUTING.md#cs) 
for this project.

-   [&#10004; ] My code follows the [skeleton code structure](https://github.com/FOSS-UCSC/FOSSALGO/blob/master/CONTRIBUTING.md#sample).

-   [&#10004; ] This pull request has a descriptive title. 
For example, `Added {Algorithm/DS name} [{Language}]`, not `Update README.md` or `Added new code`.

-   [&#10004; ] This pull request shall only be merged once the build passes.

-   [&#10004; ] I have mentioned the issue number correctly (with hyper-link) in this pull request description.

### CODE REVIEW

File: algorithms/gr-dfsrh/python3/dfsrh.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+	def DFSUtil(self, v, visited): 
+
+		# Mark the current node as visited 
+		# and print it 
+		visited[v] = True
+		print(v, end = ' ') 
+
+		# Recur for all the vertices 
+		# adjacent to this vertex 
+		for i in self.graph[v]:
```
Comment (RavinduSachintha) [inline algorithms/gr-dfsrh/python3/dfsrh.py:29]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6323759)

File: algorithms/gr-dfsrh/python3/dfsrh.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+	# Constructor 
+	def __init__(self): 
+
+		# default dictionary to store graph 
+		self.graph = defaultdict(list) 
+
+	# function to add an edge to graph 
+	def addEdge(self, u, v): 
+		self.graph[u].append(v)
```
Comment (RavinduSachintha) [inline algorithms/gr-dfsrh/python3/dfsrh.py:17]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6323759)

File: algorithms/gr-dfsrh/python3/dfsrh.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
+# Python3 program to print DFS traversal 
+# from a given given graph 
+from collections import defaultdict 
+
+# This class represents a directed graph using 
+# adjacency list representation 
+class Graph:
```
Comment (RavinduSachintha) [inline algorithms/gr-dfsrh/python3/dfsrh.py:7]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6323759)

File: algorithms/gr-dfsrh/python3/dfsrh.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+	# A function used by DFS 
+	def DFSUtil(self, v, visited): 
+
+		# Mark the current node as visited 
+		# and print it 
+		visited[v] = True
+		print(v, end = ' ') 
+
+		# Recur for all the vertices
```
Comment (RavinduSachintha) [inline algorithms/gr-dfsrh/python3/dfsrh.py:27]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6323759)

File: algorithms/gr-dfsrh/python3/dfsrh.py [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+				self.DFSUtil(i, visited) 
+
+	# The function to do DFS traversal. It uses 
+	# recursive DFSUtil() 
+	def DFS(self, v): 
+
+		# Mark all the vertices as not visited 
+		visited = [False] * (max(self.graph)+1) 
+
+		# Call the recursive helper function
```
Comment (RavinduSachintha) [inline algorithms/gr-dfsrh/python3/dfsrh.py:40]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6323759)

File: algorithms/gr-dfsrh/python3/dfsrh.py [hunk 6]
```diff
@@ -0,0 +1,59 @@
+# Python3 program to print DFS traversal 
+# from a given given graph 
+from collections import defaultdict 
+
+# This class represents a directed graph using
```
Comment (RavinduSachintha) [inline algorithms/gr-dfsrh/python3/dfsrh.py:5]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6323759)

File: algorithms/gr-dfsrh/python3/dfsrh.py [hunk 7]
```diff
@@ -0,0 +1,59 @@
+# Python3 program to print DFS traversal
```
Comment (RavinduSachintha) [inline algorithms/gr-dfsrh/python3/dfsrh.py:1]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6323759)

File: algorithms/gr-dfsrh/python3/dfsrh.py [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+
+		# Mark the current node as visited 
+		# and print it 
+		visited[v] = True
+		print(v, end = ' ') 
+
+		# Recur for all the vertices 
+		# adjacent to this vertex 
+		for i in self.graph[v]: 
+			if visited[i] == False:
```
Comment (RavinduSachintha) [inline algorithms/gr-dfsrh/python3/dfsrh.py:30]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6323759)

File: algorithms/gr-dfsrh/python3/dfsrh.py [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+		visited[v] = True
+		print(v, end = ' ') 
+
+		# Recur for all the vertices 
+		# adjacent to this vertex 
+		for i in self.graph[v]: 
+			if visited[i] == False: 
+				self.DFSUtil(i, visited) 
+
+	# The function to do DFS traversal. It uses
```
Comment (RavinduSachintha) [inline algorithms/gr-dfsrh/python3/dfsrh.py:33]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6323759)

File: algorithms/gr-dfsrh/python3/dfsrh.py [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+# Driver code 
+
+# Create a graph given 
+# in the above diagram 
+g = Graph() 
+g.addEdge(0, 1) 
+g.addEdge(0, 2) 
+g.addEdge(1, 2) 
+g.addEdge(2, 0) 
+g.addEdge(2, 3)
```
Comment (RavinduSachintha) [inline algorithms/gr-dfsrh/python3/dfsrh.py:53]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6323759)

File: algorithms/gr-dfsrh/python3/dfsrh.py [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+	# A function used by DFS 
+	def DFSUtil(self, v, visited): 
+
+		# Mark the current node as visited 
+		# and print it 
+		visited[v] = True
+		print(v, end = ' ') 
+
+		# Recur for all the vertices 
+		# adjacent to this vertex
```
Comment (RavinduSachintha) [inline algorithms/gr-dfsrh/python3/dfsrh.py:28]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6323759)

File: algorithms/gr-dfsrh/python3/dfsrh.py [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+
+		# Recur for all the vertices 
+		# adjacent to this vertex 
+		for i in self.graph[v]: 
+			if visited[i] == False: 
+				self.DFSUtil(i, visited) 
+
+	# The function to do DFS traversal. It uses 
+	# recursive DFSUtil() 
+	def DFS(self, v):
```
Comment (RavinduSachintha) [inline algorithms/gr-dfsrh/python3/dfsrh.py:35]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6323759)

File: algorithms/gr-dfsrh/python3/dfsrh.py [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+g.addEdge(0, 2) 
+g.addEdge(1, 2) 
+g.addEdge(2, 0) 
+g.addEdge(2, 3) 
+g.addEdge(3, 3) 
+
+print("Following is DFS from (starting from vertex 2)") 
+g.DFS(2) 
+
+# This code is contributed by Neelam Yadav
```
Comment (RavinduSachintha) [inline algorithms/gr-dfsrh/python3/dfsrh.py:59]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6323759)

File: algorithms/gr-dfsrh/python3/dfsrh.py [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+		# adjacent to this vertex 
+		for i in self.graph[v]: 
+			if visited[i] == False: 
+				self.DFSUtil(i, visited) 
+
+	# The function to do DFS traversal. It uses 
+	# recursive DFSUtil() 
+	def DFS(self, v): 
+
+		# Mark all the vertices as not visited
```
Comment (RavinduSachintha) [inline algorithms/gr-dfsrh/python3/dfsrh.py:37]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6323759)

File: algorithms/gr-dfsrh/python3/dfsrh.py [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+		# Call the recursive helper function 
+		# to print DFS traversal 
+		self.DFSUtil(v, visited) 
+
+# Driver code 
+
+# Create a graph given 
+# in the above diagram 
+g = Graph() 
+g.addEdge(0, 1)
```
Comment (RavinduSachintha) [inline algorithms/gr-dfsrh/python3/dfsrh.py:49]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6323759)

File: algorithms/gr-dfsrh/python3/dfsrh.py [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+
+		# Call the recursive helper function 
+		# to print DFS traversal 
+		self.DFSUtil(v, visited) 
+
+# Driver code 
+
+# Create a graph given 
+# in the above diagram 
+g = Graph()
```
Comment (RavinduSachintha) [inline algorithms/gr-dfsrh/python3/dfsrh.py:48]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6323759)

File: algorithms/gr-dfsrh/python3/dfsrh.py [hunk 17]
```diff
@@ -0,0 +1,59 @@
...
+from collections import defaultdict 
+
+# This class represents a directed graph using 
+# adjacency list representation 
+class Graph: 
+
+	# Constructor 
+	def __init__(self): 
+
+		# default dictionary to store graph
```
Comment (RavinduSachintha) [inline algorithms/gr-dfsrh/python3/dfsrh.py:12]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6323759)

File: algorithms/gr-dfsrh/python3/dfsrh.py [hunk 18]
```diff
@@ -0,0 +1,59 @@
...
+
+	# function to add an edge to graph 
+	def addEdge(self, u, v): 
+		self.graph[u].append(v) 
+
+	# A function used by DFS 
+	def DFSUtil(self, v, visited): 
+
+		# Mark the current node as visited 
+		# and print it
```
Comment (RavinduSachintha) [inline algorithms/gr-dfsrh/python3/dfsrh.py:23]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6323759)

File: algorithms/gr-dfsrh/python3/dfsrh.py [hunk 19]
```diff
@@ -0,0 +1,59 @@
...
+# adjacency list representation 
+class Graph: 
+
+	# Constructor 
+	def __init__(self): 
+
+		# default dictionary to store graph 
+		self.graph = defaultdict(list) 
+
+	# function to add an edge to graph
```
Comment (RavinduSachintha) [inline algorithms/gr-dfsrh/python3/dfsrh.py:15]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6323759)

File: algorithms/gr-dfsrh/python3/dfsrh.py [hunk 20]
```diff
@@ -0,0 +1,59 @@
...
+	def DFS(self, v): 
+
+		# Mark all the vertices as not visited 
+		visited = [False] * (max(self.graph)+1) 
+
+		# Call the recursive helper function 
+		# to print DFS traversal 
+		self.DFSUtil(v, visited) 
+
+# Driver code
```
Comment (RavinduSachintha) [inline algorithms/gr-dfsrh/python3/dfsrh.py:44]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6323759)

File: algorithms/gr-dfsrh/python3/dfsrh.py [hunk 21]
```diff
@@ -0,0 +1,59 @@
...
+
+# This class represents a directed graph using 
+# adjacency list representation 
+class Graph: 
+
+	# Constructor 
+	def __init__(self): 
+
+		# default dictionary to store graph 
+		self.graph = defaultdict(list)
```
Comment (RavinduSachintha) [inline algorithms/gr-dfsrh/python3/dfsrh.py:13]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6323759)

File: algorithms/gr-dfsrh/python3/dfsrh.py [hunk 22]
```diff
@@ -0,0 +1,59 @@
...
+class Graph: 
+
+	# Constructor 
+	def __init__(self): 
+
+		# default dictionary to store graph 
+		self.graph = defaultdict(list) 
+
+	# function to add an edge to graph 
+	def addEdge(self, u, v):
```
Comment (RavinduSachintha) [inline algorithms/gr-dfsrh/python3/dfsrh.py:16]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6323759)

File: algorithms/gr-dfsrh/python3/dfsrh.py [hunk 23]
```diff
@@ -0,0 +1,59 @@
+# Python3 program to print DFS traversal 
+# from a given given graph 
+from collections import defaultdict 
+
+# This class represents a directed graph using 
+# adjacency list representation 
+class Graph: 
+
+	# Constructor 
+	def __init__(self):
```
Comment (RavinduSachintha) [inline algorithms/gr-dfsrh/python3/dfsrh.py:10]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6323759)

File: algorithms/gr-dfsrh/python3/dfsrh.py [hunk 24]
```diff
@@ -0,0 +1,59 @@
...
+g = Graph() 
+g.addEdge(0, 1) 
+g.addEdge(0, 2) 
+g.addEdge(1, 2) 
+g.addEdge(2, 0) 
+g.addEdge(2, 3) 
+g.addEdge(3, 3) 
+
+print("Following is DFS from (starting from vertex 2)") 
+g.DFS(2)
```
Comment (RavinduSachintha) [inline algorithms/gr-dfsrh/python3/dfsrh.py:57]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6323759)

File: algorithms/gr-dfsrh/python3/dfsrh.py [hunk 25]
```diff
@@ -0,0 +1,59 @@
...
+		self.DFSUtil(v, visited) 
+
+# Driver code 
+
+# Create a graph given 
+# in the above diagram 
+g = Graph() 
+g.addEdge(0, 1) 
+g.addEdge(0, 2) 
+g.addEdge(1, 2)
```
Comment (RavinduSachintha) [inline algorithms/gr-dfsrh/python3/dfsrh.py:51]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6323759)

**Review by Danushka96 — APPROVED** [review]

**Pull request merged**